rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,296 @@
+that is lower than the static size in the other type are not compatible. The
+above example type is compatible with `tensor<2x5xf32>` but not with
+`tensor<7x5xf32>` as it doesn’t respect the bound `3` on the first dimension.
+
+Currently, the StableHLO dialect is using the MLIR core ranked tensor type to
+represent bounds. It should be noted that there is a plan to introduce a custom
+StableHLO type in the future that could natively support bounds along with
+custom pretty printing format. There will be a separate RFC on this. Also, the
+proposal will follow StableHLO backward compatibility policies so it is safe to
+use this type currently.
+
+
+## (P2) StableHLO op semantics with bounded operands or results
+
+All ops that support dynamic operands or results can have bounds specified for
+them. However, the result types need to be compatible with the inferred result
+types. This allows result types to be more generic or specific as long as it is
+compatible with the inferred type.
+
+Separately, StableHLO specification will be updated to cover bounded types for
+all the relevant ops.
+
+## (P3) get\_dimension\_size / set\_dimension\_size ops
+
+The `get_dimension_size` op takes a tensor and a dimension index and returns the
+actual size of the dimension at runtime as an `i32` type scalar.
+
+The following example returns the size of the result after concatenating input
+that has up to `16` elements with self and returns the actual runtime size of
+the concatenation result.
+
+```
+func.func @self_concat_size(%data: tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>) -> tensor<i32> {
+  %concat = "stablehlo.concatenate"(%data, %data) {dimension = 0 : i64}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>,
+       tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>)
+      -> tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>
+
+  %result = stablehlo.get_dimension_size %concat, dim = 0
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>) -> tensor<i32>
+
+  func.return %result : tensor<i32>
+}
+```
+
+The `set_dimension_size` op takes a bounded tensor, runtime size and a dimension
+index and returns a tensor whose logical size of the particular dimension is set
+to the specified size. This size needs to be less than or equal to the static
+size or bound for the dimension. This operation can be thought as either a slice
+or pad operation depending on if the earlier logical dimension size is larger or
+smaller, respectively. In case the dimension size is increased, the padded
+values are undefined.
+
+In the following example, `set_dimension_size` op is used to set the logical
+size of the first dimension so that it performs a sum reduction on the first
+`batch_size` elements in the input. With data argument `[1, 2, 3, 4]` and
+batch\_size argument `2`, the following function returns `3` but it returns `6`
+for the same data argument when the batch\_size is `3`. Operand static dimension
```

Comment:
This sentence is hard to read. Did you mean something like

The `set_dimension_size` op also sets the static bound of the returned tensor by providing a bound in the result type. In this example, the bound is set to `4`.

Reply:
Thanks! Rephrased it.